### PR TITLE
wasm: provide working writev() syscall.

### PR DESCRIPTION
**Pull request by PiotrSikora**
Implementation restricted to stdout/stderr redirected to Envoy logs.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

### CODE REVIEW

**Comment by PiotrSikora** [comment]
@jplevyak it doesn't appear that `__syscall4` is imported by any WASM modules, so I'd skip it for now.

**Comment by jplevyak** [comment]
Yes, it looks like write() is the only function which calls SYS_write.

File: source/extensions/common/wasm/wasm.cc [hunk 1]
```diff
@@ -535,7 +535,48 @@ Word ___syscall54Handler(void*, Word, Word) { throw WasmException("emscripten sy
...
+    const uint32_t* iovec = reinterpret_cast<const uint32_t*>(memslice.data());
+    if (iovec[1] /* size */) {
+      memslice = context->wasmVm()->getMemory(iovec[0] /* data */, iovec[1] /* size */);
+      s.append(memslice.data(), memslice.size());
+    }
+  }
+
+  size_t written = s.size();
+  if (written) {
+    // Remove trailing newline from the logs, if any.
```
Comment (silentdai) [inline source/extensions/common/wasm/wasm.cc]: I would be surprised when I get the same log despite I write with "\n" or without "\n". TBH I don't think it is the syscall layer to handle this logic.

Comment (PiotrSikora) [reply source/extensions/common/wasm/wasm.cc]: We're intercepting writes to stdout/stderr and writing them to Envoy logs, where each entry already has a newline at the end. If we don't remove those trailing newlines, then each intercepted write to stdout/stderr will be followed by an empty line.

I don't think that removing empty lines is more intrusive than prefixing those writes with a timestamp, filename and `wasm log:`, so I'm going to leave this as-is.

**Pull request merged**